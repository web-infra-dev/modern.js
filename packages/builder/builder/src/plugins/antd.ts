import { getAntdMajorVersion } from '@modern-js/utils';
import type {
  BuilderTarget,
  DefaultBuilderPlugin,
} from '@modern-js/builder-shared';
import { SourceConfig as WebpackSourceConfig } from '@modern-js/builder-webpack-provider';
import { SourceConfig as RspackSourceConfig } from '@modern-js/builder-rspack-provider';

export const builderPluginAntd = (): DefaultBuilderPlugin => ({
  name: `builder-plugin-antd`,

  setup(api) {
    api.modifyBuilderConfig(builderConfig => {
      builderConfig.source ??= {};

      const { transformImport } = builderConfig.source as RspackSourceConfig &
        WebpackSourceConfig;

      if (
        transformImport === false ||
        transformImport?.some(item => item.libraryName === 'antd')
      ) {
        return;
      }

      const antdMajorVersion = getAntdMajorVersion(api.context.rootPath);
      // antd >= v5 no longer need babel-plugin-import
      // see: https://ant.design/docs/react/migration-v5#remove-babel-plugin-import
      if (antdMajorVersion && antdMajorVersion < 5) {
        (builderConfig.source as WebpackSourceConfig).transformImport = [
          ...(transformImport || []),
          {
            libraryName: 'antd',
            libraryDirectory: useSSR(api.context.target) ? 'lib' : 'es',
            style: true,
          },
        ];
      }
    });
  },
});

export function useSSR(target: BuilderTarget | BuilderTarget[]) {
  return (Array.isArray(target) ? target : [target]).some(item =>
    ['node', 'service-worker'].includes(item),
  );
}
