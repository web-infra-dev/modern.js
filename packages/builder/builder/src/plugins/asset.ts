import { join } from 'path';
import {
  getRegExpForExts,
  getDistPath,
  getFilename,
  chainStaticAssetRule,
  FilenameConfig,
} from '@modern-js/builder-shared';
import type { DefaultBuilderPlugin } from '@modern-js/builder-shared';

export const builderAssetPlugin = (
  assetType: keyof FilenameConfig,
  exts: string[],
): DefaultBuilderPlugin => ({
  name: `builder-plugin-${assetType}`,

  setup(api) {
    api.modifyBundlerChain((chain, { isProd }) => {
      const config = api.getNormalizedConfig();
      const regExp = getRegExpForExts(exts);
      const distDir = getDistPath(config.output, assetType);
      const filename = getFilename(config.output, assetType, isProd);

      const maxSize = config.output.dataUriLimit.image;

      const rule = chain.module.rule(assetType).test(regExp);

      chainStaticAssetRule({
        rule,
        maxSize,
        filename: join(distDir, filename),
        assetType,
      });
    });
  },
});
