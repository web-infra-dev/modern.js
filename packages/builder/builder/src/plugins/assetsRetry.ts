import {
  getDistPath,
  isHtmlDisabled,
  DefaultBuilderPlugin,
} from '@modern-js/builder-shared';

export function builderPluginAssetsRetry(): DefaultBuilderPlugin {
  return {
    name: 'builder-plugin-assets-retry',
    setup(api) {
      api.modifyBundlerChain(
        async (chain, { CHAIN_ID, target, HtmlPlugin }) => {
          const config = api.getNormalizedConfig();

          if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
            return;
          }

          const { AssetsRetryPlugin } = await import(
            '@modern-js/builder-shared'
          );
          const distDir = getDistPath(config.output, 'js');

          const { assetsRetry = {} } = config.output;

          // assetsRetry.crossOrigin should be same as html.crossorigin by default
          if (assetsRetry.crossOrigin === undefined) {
            assetsRetry.crossOrigin = config.html.crossorigin;
          }

          chain.plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY).use(AssetsRetryPlugin, [
            {
              ...assetsRetry,
              distDir,
              HtmlPlugin,
            },
          ]);
        },
      );
    },
  };
}
