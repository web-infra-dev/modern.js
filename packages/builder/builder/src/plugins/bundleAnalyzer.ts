import { DefaultBuilderPlugin } from '@modern-js/builder-shared';

export function builderPluginBundleAnalyzer(): DefaultBuilderPlugin {
  return {
    name: 'builder-plugin-bundle-analyzer',
    setup(api) {
      api.modifyBundlerChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();
        // There are two ways to open the bundle analyzer:
        // 1. Set environment variable `BUNDLE_ANALYZE`
        // 2. Set performance.bundleAnalyze config
        if (!process.env.BUNDLE_ANALYZE && !config.performance.bundleAnalyze) {
          return;
        }
        const { default: BundleAnalyzer } = await import(
          '@modern-js/builder-shared/webpack-bundle-analyzer'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.BUNDLE_ANALYZER)
          .use(BundleAnalyzer.BundleAnalyzerPlugin, [
            {
              analyzerMode: 'static',
              openAnalyzer: false,
              reportFilename: `report-${target}.html`,
              ...(config.performance.bundleAnalyze || {}),
            },
          ]);
      });
    },
  };
}
