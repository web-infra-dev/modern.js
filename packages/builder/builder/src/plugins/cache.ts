import { isAbsolute, join } from 'path';
import {
  BuildCacheOptions,
  BuilderContext,
  isFileExists,
  DefaultBuilderPlugin,
} from '@modern-js/builder-shared';

async function validateCache(
  cacheDirectory: string,
  buildDependencies: Record<string, string[]>,
) {
  const { fs } = await import('@modern-js/utils');
  const configFile = join(cacheDirectory, 'buildDependencies.json');

  if (await isFileExists(configFile)) {
    const prevBuildDependencies = await fs.readJSON(configFile);

    if (
      JSON.stringify(prevBuildDependencies) ===
      JSON.stringify(buildDependencies)
    ) {
      return;
    }

    /**
     * If the filenames in the buildDependencies are changed, webpack will not invalidate the previous cache.
     * So we need to remove the cache directory to make sure the cache is invalidated.
     */
    await fs.remove(cacheDirectory);
  }

  await fs.outputJSON(configFile, buildDependencies);
}

function getCacheDirectory(
  { cacheDirectory }: BuildCacheOptions,
  context: BuilderContext,
) {
  if (cacheDirectory) {
    return isAbsolute(cacheDirectory)
      ? cacheDirectory
      : join(context.rootPath, cacheDirectory);
  }
  return join(context.cachePath, context.bundlerType);
}

export const PluginCache = (): DefaultBuilderPlugin => ({
  name: 'builder-plugin-cache',

  setup(api) {
    api.modifyBundlerChain(async (chain, { target, env }) => {
      const { buildCache } = api.getNormalizedConfig().performance;

      if (buildCache === false) {
        chain.cache(false);
        return;
      }

      const { context } = api;
      const cacheConfig = typeof buildCache === 'boolean' ? {} : buildCache;

      const cacheDirectory = getCacheDirectory(cacheConfig, context);
      const rootPackageJson = join(context.rootPath, 'package.json');
      const browserslistConfig = join(context.rootPath, '.browserslistrc');

      /**
       * webpack can't detect the changes of framework config, tsconfig and browserslist config.
       * but they will affect the compilation result, so they need to be added to buildDependencies.
       */
      const buildDependencies: Record<string, string[]> = {
        packageJson: [rootPackageJson],
      };
      if (context.configPath) {
        buildDependencies.config = [context.configPath];
      }
      if (context.tsconfigPath) {
        buildDependencies.tsconfig = [context.tsconfigPath];
      }
      if (await isFileExists(browserslistConfig)) {
        buildDependencies.browserslistrc = [browserslistConfig];
      }

      await validateCache(cacheDirectory, buildDependencies);

      chain.cache({
        // The default cache name of webpack is '${name}-${env}', and the `name` is `default` by default.
        // We set cache name to avoid cache conflicts of different targets.
        name: `${target}-${env}`,
        type: 'filesystem',
        cacheDirectory,
        buildDependencies,
      });
    });
  },
});
