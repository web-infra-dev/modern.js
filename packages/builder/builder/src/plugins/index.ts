export const plugins = {
  cleanOutput: () => import('./cleanOutput').then(m => m.PluginCleanOutput()),
  startUrl: () => import('./startUrl').then(m => m.PluginStartUrl()),
  fileSize: () => import('./fileSize').then(m => m.PluginFileSize()),
  devtool: () => import('./devtool').then(m => m.PluginDevtool()),
  target: () => import('./target').then(m => m.PluginTarget()),
  entry: () => import('./entry').then(m => m.PluginEntry()),
};
