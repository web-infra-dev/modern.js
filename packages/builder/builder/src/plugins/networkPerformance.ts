import {
  HtmlNetworkPerformancePlugin,
  type DefaultBuilderPlugin,
} from '@modern-js/builder-shared';

export const builderPluginNetworkPerformance = (): DefaultBuilderPlugin => ({
  name: `builder-plugin-network-performance`,

  setup(api) {
    api.modifyBundlerChain(
      async (
        chain,
        { CHAIN_ID, isServer, isWebWorker, isServiceWorker, HtmlPlugin },
      ) => {
        const config = api.getNormalizedConfig();
        const {
          performance: { dnsPrefetch, preconnect },
        } = config;

        if (isServer || isWebWorker || isServiceWorker) {
          return;
        }

        if (dnsPrefetch) {
          chain
            .plugin(CHAIN_ID.PLUGIN.HTML_DNS_PREFETCH)
            .use(HtmlNetworkPerformancePlugin, [
              dnsPrefetch,
              'dnsPrefetch',
              HtmlPlugin,
            ]);
        }

        if (preconnect) {
          chain
            .plugin(CHAIN_ID.PLUGIN.HTML_PRECONNECT)
            .use(HtmlNetworkPerformancePlugin, [
              preconnect,
              'preconnect',
              HtmlPlugin,
            ]);
        }
      },
    );
  },
});
