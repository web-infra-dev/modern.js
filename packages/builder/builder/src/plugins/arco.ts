import { isPackageInstalled } from '@modern-js/utils';
import { useSSR } from './antd';
import type { DefaultBuilderPlugin } from '@modern-js/builder-shared';
import { SourceConfig as WebpackSourceConfig } from '@modern-js/builder-webpack-provider';
import { SourceConfig as RspackSourceConfig } from '@modern-js/builder-rspack-provider';

export const builderPluginArco = (): DefaultBuilderPlugin => ({
  name: `builder-plugin-arco`,

  setup(api) {
    const ARCO_NAME = '@arco-design/web-react';
    const ARCO_ICON = `${ARCO_NAME}/icon`;

    api.modifyBuilderConfig(builderConfig => {
      const { transformImport = [] } = (builderConfig.source ||
        {}) as WebpackSourceConfig & RspackSourceConfig;

      if (
        transformImport === false ||
        !isPackageInstalled(ARCO_NAME, api.context.rootPath)
      ) {
        return;
      }

      const isUseSSR = useSSR(api.context.target);

      if (!transformImport?.some(item => item.libraryName === ARCO_NAME)) {
        transformImport.push({
          libraryName: ARCO_NAME,
          libraryDirectory: isUseSSR ? 'lib' : 'es',
          camelToDashComponentName: false,
          style: true,
        });
      }

      if (!transformImport?.some(item => item.libraryName === ARCO_ICON)) {
        transformImport.push({
          libraryName: ARCO_ICON,
          libraryDirectory: isUseSSR ? 'react-icon-cjs' : 'react-icon',
          camelToDashComponentName: false,
        });
      }

      builderConfig.source ||= {};
      (
        builderConfig.source as RspackSourceConfig & WebpackSourceConfig
      ).transformImport = transformImport;
    });
  },
});
