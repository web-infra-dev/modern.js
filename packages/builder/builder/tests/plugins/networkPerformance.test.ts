import { expect, describe, it } from 'vitest';
import * as shared from '@modern-js/builder-shared';
import { builderPluginNetworkPerformance } from '@/plugins/networkPerformance';

describe('plugins/builderPluginNetworkPerformance', () => {
  it('should add network performance plugin', async () => {
    let modifyBundlerChainCb: any;

    const api: any = {
      modifyBundlerChain: (fn: any) => {
        modifyBundlerChainCb = fn;
      },
      getNormalizedConfig: () => ({
        performance: {
          preconnect: 'http://aaa.com',
          dnsPrefetch: 'http://aaa.com',
        },
      }),
      context: {},
    };
    builderPluginNetworkPerformance().setup(api);

    const chain = await shared.getBundlerChain();

    await modifyBundlerChainCb(chain, {
      CHAIN_ID: {
        PLUGIN: {
          HTML_PRECONNECT: 'HTML_PRECONNECT',
          HTML_DNS_PREFETCH: 'HTML_DNS_PREFETCH',
        },
      },
      HtmlPlugin: {
        name: 'HtmlPlugin',
      },
      target: 'web',
    });

    expect(chain.toConfig()).toMatchSnapshot();
  });

  it('should not add network performance plugin when target is server', async () => {
    let modifyBundlerChainCb: any;

    const api: any = {
      modifyBundlerChain: (fn: any) => {
        modifyBundlerChainCb = fn;
      },
      getNormalizedConfig: () => ({
        performance: {
          preconnect: 'http://aaa.com',
          dnsPrefetch: 'http://aaa.com',
        },
      }),
      context: {},
    };
    builderPluginNetworkPerformance().setup(api);

    const chain = await shared.getBundlerChain();

    await modifyBundlerChainCb(chain, {
      CHAIN_ID: {
        PLUGIN: {
          HTML_PRECONNECT: 'HTML_PRECONNECT',
          HTML_DNS_PREFETCH: 'HTML_DNS_PREFETCH',
        },
      },
      HtmlPlugin: {
        name: 'HtmlPlugin',
      },
      isServer: true,
      target: 'web',
    });

    expect(chain.toConfig()).toMatchInlineSnapshot('{}');
  });

  it('should not add network performance plugin by default', async () => {
    let modifyBundlerChainCb: any;

    const api: any = {
      modifyBundlerChain: (fn: any) => {
        modifyBundlerChainCb = fn;
      },
      getNormalizedConfig: () => ({
        performance: {},
      }),
      context: {},
    };
    builderPluginNetworkPerformance().setup(api);

    const chain = await shared.getBundlerChain();

    await modifyBundlerChainCb(chain, {
      CHAIN_ID: {
        PLUGIN: {
          HTML_PRECONNECT: 'HTML_PRECONNECT',
          HTML_DNS_PREFETCH: 'HTML_DNS_PREFETCH',
        },
      },
      HtmlPlugin: {
        name: 'HtmlPlugin',
      },
      isServer: true,
      target: 'web',
    });

    expect(chain.toConfig()).toMatchInlineSnapshot('{}');
  });
});
