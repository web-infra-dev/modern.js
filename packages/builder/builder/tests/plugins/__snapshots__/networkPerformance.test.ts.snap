// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/builderPluginNetworkPerformance > should add network performance plugin 1`] = `
{
  "plugins": [
    HtmlNetworkPerformancePlugin {
      "HtmlPlugin": {
        "name": "HtmlPlugin",
      },
      "options": "/http/aaa.com",
      "type": "dnsPrefetch",
    },
    HtmlNetworkPerformancePlugin {
      "HtmlPlugin": {
        "name": "HtmlPlugin",
      },
      "options": "/http/aaa.com",
      "type": "preconnect",
    },
  ],
}
`;
